akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -133,6 +135,167 @@ import scala.util.control.NonFatal
+          pull(in)
+        }
+        else if(accumulated.size == n) { //corner case for n = 0, can be handled in FlowOps
+          materializeFlow()
+        }
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = {
+        if(null != subSink){
+          subSink.cancel(cause)
+        } else {
+          materializeFlow()
+        }
+      }
+
+      def materializeFlow(): Unit = {
+        val prefix = accumulated.reverse
+        accumulated = Nil
+        val flow = try f(prefix) catch {
+          case NonFatal(ex) =>
+            val neverMaterializedEx = new NeverMaterializedException(ex)
+            matPromise.failure(neverMaterializedEx)
+            throw neverMaterializedEx
+        }
+        subSource = new SubSourceOutlet[In]("subSource")
+        subSource.setHandler{
+          new OutHandler {
+            /**
+             * Called when the output port has received a pull, and therefore ready to emit an element, i.e. [[GraphStageLogic.push]]
+             * is now allowed to be called on this port.
+             */
+            override def onPull(): Unit = {
+              if(!isClosed(in) && !hasBeenPulled(in)) {
+                pull(in)
+              }
+            }
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              if(!isClosed(in)) {
+                cancel(in, cause)
+              }
+              super.onDownstreamFinish(cause)
+            }
+          }
+        }
+        subSink = new SubSinkInlet[Out]("subSink")
+        subSink.setHandler{
+          new InHandler {
+            /**
+             * Called when the input port has a new element available. The actual element can be retrieved via the
+             * [[GraphStageLogic.grab]] method.
+             */
+            override def onPush(): Unit = {
+              push(out, subSink.grab())
+            }
+
+            override def onUpstreamFinish(): Unit = {
+              completeStage()
+              super.onUpstreamFinish()
+            }
+
+            override def onUpstreamFailure(ex: Throwable): Unit = {
+              failStage(ex)
+              super.onUpstreamFailure(ex)
+            }
+          }
+        }
+        val runnableGraph = Source.fromGraph(subSource.source)
+          .viaMat(flow)(Keep.right)
+          .to(subSink.sink)
+        val m = Try(interpreter.subFusingMaterializer.materialize(runnableGraph))
+        matPromise.complete(m)
+        m.get
+        //this materialization comes in response to downstream demand, so we need to signal this to the materialized flow
+        if(!isClosed(in)) {
+          subSink.pull()
+        } else {
+          subSource.complete()
+        }
+      }
+    }
+    (logic, matPromise.future)
+  }
+
+  /**
+   * The shape of a graph is all that is externally visible: its inlets and outlets.
+   */
+  override val shape: FlowShape[In, Out] = {
+    FlowShape.of(Inlet[In]("in"), Outlet[Out]("out"))
+  }
+}
+
```

Comment:
I know we have dumped all kinds of operators in two or three huge file previously but please put this operator in a separate file to keep things neat rather than pile up on that

Reply:
👍 will do